src/hotspot/share/gc/shenandoah/shenandoahConcurrentMark.cpp

```diff
@@ -216,14 +218,20 @@ void ShenandoahConcurrentMark::concurrent_mark() {
     }
 
     size_t before = qset.completed_buffers_num();
+    double handshake_start_time = os::elapsedTime();
     Handshake::execute(&flush_satb);
+    handshake_total_time += os::elapsedTime() - handshake_start_time;
     size_t after = qset.completed_buffers_num();
 
     if (before == after) {
       // No more retries needed, break out.
       break;
     }
   }
+  ShenandoahPhaseTimings* const timings = heap->phase_timings();
+  timings->record_phase_time(ShenandoahPhaseTimings::conc_mark_satb_flush_rendezvous, handshake_total_time);
+
+  log_info(gc, phases) ("Total SATB flush attempts : %d", (flushes + 1));
```

Comment:
Instead of this, I suggest we wrap the `Handshake::execute` with `log_info(gc,phases,start)` and `log_info(gc,phases)`. Logging would tell us how many attempts were done, and give us a rough number how long it took. The benefit would be that we would have the exact timestamps when handshakes started/finished.

Reply:
Hm. Maybe even better to wrap it with:

```
    {
      ShenandoahTimingsTracker t(ShenandoahPhaseTimings::conc_mark_satb_flush_rendezvous, true);
      Handshake::execute(&flush_satb);
    }
```

Note the last, new argument that allows to merge multiple events into one. We need to separately make `ShenandoahTimingsTracker` report `gc,phases` log statement, imo.

Reply:
Do we want something like below for conc_mark_satb_flush_rendezvous phase?
The log line is not useful for all the other usage of ShenandoahTimingTracker. So I'm thinking of logging only when boolean should_aggregate is set.

There will be two log lines when number of flush attempts = 2.

```
[3.475s][info][gc,phases   ] GC(14) Recorded time for phase SATB Flush Rendezvous as 0.615854 ms
[3.476s][info][gc,phases   ] GC(14) Recorded time for phase SATB Flush Rendezvous as 0.357069 ms
```
We can also have separate log line for start and end.

Reply:
I don't think we need this phase tracking to do anything special, really. What we should probably do is to feed it into `ShenandoahTimingsTracker`, and let the rest of the infra deal with it as it sees fit. Nominally, show it in `gc+stats`. Later, we would need to consider printing `gc,phases` logging straight from `ShenandoahTimingsTracker`.